src/ucp/wireup/select.c

```diff
@@ -801,16 +822,52 @@ static ucs_status_t ucp_wireup_add_rndv_lane(ucp_ep_h ep,
         criteria.local_iface_flags |= UCP_WORKER_UCT_UNSIG_EVENT_CAP_FLAGS;
     }
 
-    status = ucp_wireup_select_transport(ep, address_list, address_count, &criteria,
-                                         -1, -1, 0, &rsc_index, &addr_index, &score);
-    if ((status == UCS_OK) &&
-        /* a temporary workaround to prevent the ugni uct from using rndv */
-        (strstr(ep->worker->context->tl_rscs[rsc_index].tl_rsc.tl_name, "ugni") == NULL)) {
-         ucp_wireup_add_lane_desc(lane_descs, num_lanes_p, rsc_index, addr_index,
-                                 address_list[addr_index].md_index, score,
-                                 UCP_WIREUP_LANE_USAGE_RNDV, 0);
+    max_lanes = ucs_min(UCP_MAX_RNDV_LANES, ep->worker->context->config.ext.max_rndv_lanes);
+
+    /* clone address list to allow safe modification of list */
+    addrs = ucs_malloc(address_count * sizeof(*addrs), "address list copy");
+    if (addrs == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    memcpy(addrs, address_list, address_count * sizeof(*addrs));
+
+    /* every local & remote address should be selected only once, that is why
+     * we exclude selected address from processing on next iteration */
+    for (i = 0; (i < max_lanes) && address_count; /* no increment here */) {
+        status = ucp_wireup_select_transport(ep, addrs, address_count, &criteria,
+                                             tl_bitmap, -1, 0, &rsc_index, &addr_index, &score);
+        if ((status == UCS_OK) &&
+            /* a temporary workaround to prevent the ugni uct from using rndv */
+            (strstr(ep->worker->context->tl_rscs[rsc_index].tl_rsc.tl_name, "ugni") == NULL)) {
+             ucp_wireup_add_lane_desc(lane_descs, num_lanes_p, rsc_index, addr_index,
+                                      address_list[addr_index].md_index, score,
+                                      UCP_WIREUP_LANE_USAGE_RNDV, 0);
+             i++;
+        } else if (status != UCS_OK) {
+            break;
+        }
+
+        if (ep->worker->context->tl_rscs[rsc_index].tl_rsc.dev_type == UCT_DEVICE_TYPE_SHM) {
```

Comment:
don't we also need to check that same device is not used for more than one rndv lane?

Reply:
every found rsc_index is excluded from next processing

Reply:
true, but if we have 2 transports on same device (e.g dc ad rc) they would be different rsc index

Reply:
ok, fixed